DOCS-1592: Make PyInstaller recommended method and edit build command (#2669)
viamrobotics/docs: docs/registry/create/_index.md

   python -m PyInstaller --onefile --hidden-import="googleapiclient" --add-data src/arm/my_arm_kinematics.json:src/arm/ src/main.py
   ```

>    By default, the output directory for the packaged executable is <file>dist</file>, and the name of the executable is derived from the name of the input script (in this case, main).

Author: @andf-viam added all of this!

>    By default, the output directory for the packaged executable is <file>dist</file>, and the name of the executable is derived from the base name of the input script (in this case, main).

Author: ```suggestion
   By default, the output directory for the packaged executable is <file>dist</file>, and the name of the executable is derived from the name of the input script (in this case, main).
```

We recommend you use PyInstaller with the [`build-action` GitHub action](https://github.com/viamrobotics/build-action) which provides a simple cross-platform build setup for multiple platforms: x86, ARM Linux, and MacOS.
Follow the instructions to [Update an existing module using a GitHub action](/registry/upload/#update-an-existing-module-using-a-github-action) to add the build configuration to your machine.

With this approach, you can make a build script like the following to

> build your module, and configure the resulting executable (<file>dist/main</file>) as your module `"entrypoint"`:

Author: ```suggestion
build your module:
```

```sh { class="command-line"}
#!/bin/bash
set -e

sudo apt-get install -y python3-venv
python3 -m venv .venv
. .venv/bin/activate
pip3 install -r requirements.txt
python3 -m PyInstaller --onefile --hidden-import="googleapiclient" src/main.py
tar -czvf dist/archive.tar.gz dist/main
```

> This script automates the process of setting up a Python virtual environment, installing dependencies, packaging the Python module into a standalone executable using PyInstaller, and then compressing the resulting executable into a tarball.

Author: ```suggestion
This script automates the process of setting up a Python virtual environment on a Linux arm64 machine, installing dependencies, packaging the Python module into a standalone executable using PyInstaller, and then compressing the resulting executable into a tarball.
For more examples of build scripts see [Update an existing module using a GitHub action](/registry/upload/#update-an-existing-module-using-a-github-action).
```

   python -m PyInstaller --onefile --hidden-import="googleapiclient" --add-data src/arm/my_arm_kinematics.json:src/arm/ src/main.py
   ```

   By default, the output directory for the packaged executable is <file>dist</file>, and the name of the executable is derived from the name of the input script (in this case, main).

> We recommend you use PyInstaller with the [`build-action` GitHub action](https://github.com/viamrobotics/build-action) which provides a simple cross-platform build setup for multiple platforms: x86, ARM Linux, and MacOS.

Reviewer: platforms list: 'x86 and Arm Linux, and MacOS' (or something)

Follow the instructions to [Update an existing module using a GitHub action](/registry/upload/#update-an-existing-module-using-a-github-action) to add the build configuration to your machine.

With this approach, you can make a build script like the following to
build your module, and configure the resulting executable (<file>dist/main</file>) as your module `"entrypoint"`:

```sh { class="command-line"}
#!/bin/bash
set -e

sudo apt-get install -y python3-venv
python3 -m venv .venv
. .venv/bin/activate
pip3 install -r requirements.txt
python3 -m PyInstaller --onefile --hidden-import="googleapiclient" src/main.py
tar -czvf dist/archive.tar.gz dist/main
```

This script automates the process of setting up a Python virtual environment on a Linux arm64 machine, installing dependencies, packaging the Python module into a standalone executable using PyInstaller, and then compressing the resulting executable into a tarball.
For more examples of build scripts see [Update an existing module using a GitHub action](/registry/upload/#update-an-existing-module-using-a-github-action).

{{% alert title="Note" color="note" %}}

> Used in this manner, PyInstaller does not support relative imports (imports starting with `.`).

Reviewer: inability to do relative imports could alienate people who are heavy python users

consider changing to something like:

```suggestion
Note that PyInstaller does not support relative imports in entrypoints (imports starting with `.`). If you get 'ImportError: attempted relative import with no known parent package', set up a stub entrypoint as described [here](https://github.com/pyinstaller/pyinstaller/issues/2560).
```

Author: awesome thanks!

In addition, PyInstaller does not support cross-compiling: you must compile your module on the target architecture you wish to support.

> For example, you cannot run a module on a Linux `arm64` system if you compiled it using PyInstaller on a Linux `amd64` system.

Reviewer 2: Wondering if we should rephrase and emphasize using our built system.

"Viam makes this easy for you by providing a build system for modules. Simply follow these instructions in our CLI docs to automatically build for each system your module can support."

   python -m PyInstaller --onefile --hidden-import="googleapiclient" --add-data src/arm/my_arm_kinematics.json:src/arm/ src/main.py
   ```

   By default, the output directory for the packaged executable is <file>dist</file>, and the name of the executable is derived from the name of the input script (in this case, main).

> We recommend you use PyInstaller with the [`build-action` GitHub action](https://github.com/viamrobotics/build-action) which provides a simple cross-platform build setup for multiple platforms: x86, ARM Linux, and MacOS.

Author: ```suggestion
We recommend you use PyInstaller with the [`build-action` GitHub action](https://github.com/viamrobotics/build-action) which provides a simple cross-platform build setup for multiple platforms: x86 and Arm Linux, and MacOS.
```

With this approach, you can make a build script like the following to
build your module, and configure the resulting executable (<file>dist/main</file>) as your module `"entrypoint"`:

```sh { class="command-line"}
#!/bin/bash
set -e

sudo apt-get install -y python3-venv
python3 -m venv .venv
. .venv/bin/activate
pip3 install -r requirements.txt
python3 -m PyInstaller --onefile --hidden-import="googleapiclient" src/main.py
tar -czvf dist/archive.tar.gz dist/main
```

This script automates the process of setting up a Python virtual environment on a Linux arm64 machine, installing dependencies, packaging the Python module into a standalone executable using PyInstaller, and then compressing the resulting executable into a tarball.
For more examples of build scripts see [Update an existing module using a GitHub action](/registry/upload/#update-an-existing-module-using-a-github-action).

{{% alert title="Note" color="note" %}}

Note that PyInstaller does not support relative imports in entrypoints (imports starting with `.`).

> If you get 'ImportError: attempted relative import with no known parent package', set up a stub entrypoint as described [here](https://github.com/pyinstaller/pyinstaller/issues/2560).

Author: ```suggestion
If you get 'ImportError: attempted relative import with no known parent package', set up a stub entrypoint as described on [GitHub](https://github.com/pyinstaller/pyinstaller/issues/2560).
```
